binlog.cpp

```diff
@@ -0,0 +1,235 @@
+
+timeout_error:
+    close(_file);
+    _file = -1;
+open_error:
+    free(filename);
+    return false;
+}
+
+void BinLog::stop()
+{
+    mavlink_message_t msg;
+
+    if (_file == -1) {
+        log_error("BinLog not started");
+        return;
+    }
+
+    mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg, _target_system_id, MAV_COMP_ID_ALL, MAV_REMOTE_LOG_DATA_BLOCK_STOP, 1);
+    _send_msg(&msg, _target_system_id);
+
+    if (_timeout) {
+        _mainloop->del_timeout(_timeout);
+        _timeout = nullptr;
+    }
+
+    fsync(_file);
+    close(_file);
+    _file = -1;
+    _system_id = 0;
+}
+
+int BinLog::write_msg(const struct buffer *buffer)
+{
+    const bool mavlink2 = buffer->data[0] == MAVLINK_STX;
+    uint32_t msg_id;
+    uint16_t payload_len;
+    uint8_t trimmed_zeros;
+
+    _stat.write.total++;
+    _stat.write.bytes += buffer->len;
+
+    if (mavlink2) {
+        struct mavlink_router_mavlink2_header *msg
+            = (struct mavlink_router_mavlink2_header *)buffer->data;
+        msg_id = msg->msgid;
+        payload_len = msg->payload_len;
+    } else {
+        struct mavlink_router_mavlink1_header *msg
+            = (struct mavlink_router_mavlink1_header *)buffer->data;
+        msg_id = msg->msgid;
+        payload_len = msg->payload_len;
+    }
+
+    /* Check if we are interested in this msg_id */
+    if (msg_id != MAVLINK_MSG_ID_REMOTE_LOG_DATA_BLOCK) {
+        return buffer->len;
+    }
+
+    uint8_t *payload;
+
+    // TODO do we really need this?
+    if (mavlink2) {
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink2_header);
+        trimmed_zeros = get_trimmed_zeros(buffer);
+    } else {
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink1_header);
+        trimmed_zeros = 0;
+    }
+
+    if (trimmed_zeros) {
+        mavlink_remote_log_data_block_t binlog_data;
+        memcpy(&binlog_data, payload, payload_len);
+        memset(((uint8_t *)&binlog_data) + payload_len, 0, trimmed_zeros);
+        _logging_data_process(&binlog_data);
+    } else {
+        mavlink_remote_log_data_block_t *binlog_data = (mavlink_remote_log_data_block_t *)payload;
+        _logging_data_process(binlog_data);
+    }
+
+    return buffer->len;
+}
+
+void BinLog::_send_ack(uint32_t seqno)
+{
+    mavlink_message_t msg;
+
+    // Message filled a gap, or is duplicated. just send the ack
+    if (seqno < _last_acked_seqno) {
+        mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg, _target_system_id, MAV_COMP_ID_ALL, seqno, 1);
+        _send_msg(&msg, _target_system_id);
+        return;
+    }
+
+    // TODO maybe a threshould of [n]acks to be sent?
+    // TODO send ack to source only?
+    // Send nacks regarding unseen seqno
+    for (uint32_t i = _last_acked_seqno; i < seqno; i++) {
+        mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg, _target_system_id, MAV_COMP_ID_ALL, seqno, 0);
```

Comment:
Why this block?
You send a ack to each msg received. We are missing some messages? If we are I think that we should not send a ACK for those ones.